crates/rust-analyzer/src/reload.rs

```diff
@@ -59,6 +59,19 @@ impl GlobalState {
     pub(crate) fn update_configuration(&mut self, config: Config) {
         let _p = profile::span("GlobalState::update_configuration");
         let old_config = mem::replace(&mut self.config, Arc::new(config));
+        if self.config.extra_env() != old_config.extra_env() {
+            for (old_key, old_value) in self.overwritten_env_vars.drain(..) {
+                if let Some(old_value) = old_value {
+                    env::set_var(old_key, old_value);
+                } else {
+                    env::remove_var(old_key);
+                }
+            }
+            for (new_key, new_value) in self.config.extra_env() {
+                self.overwritten_env_vars.push((new_key.clone(), std::env::var_os(new_key)));
+                std::env::set_var(new_key, new_value);
+            }
+        }
```

Comment:
This setting the env process wide, we shouldn't do that here, we should only set these for cargo processes we run

Reply:
I tried that, but `MetadataCommand` has no way to specify additional environment variables when it is run.

Reply:
There is a `cargo_command` function on `MetadataCommand` that gives you back the contructed command. So we should be able to do what `exec` does ourselves with the env injected. https://github.com/oli-obk/cargo_metadata/blob/c0e2194cfd141d770bb5844ad4f40aad8fbbd92b/src/lib.rs#L699-L711

Reply:
Ok, done.